File: archive/web-poc/public/app/audio/sleep-tone-dsp.js
import {
  FAN_AIR_BROWN_MIX,
  FAN_RUMBLE_FILTERS,
  fanAirLevel,
  fanHumFrequency,
  fanHumLevel,
  fanRumbleLevel,
} from "./tone-shaping.js";

export const DEFAULT_SLEEP_TONE_CONFIG = Object.freeze({
  fanAir: 0.55,
  fanRumble: 0.65,
  fanHum: 0.52,
  fanHumPitch: 92,
  fanDrift: 0.32,
  greenMix: 0.25,
  warmth: 0.35,
  lowCut: 0,
  highCut: 1,
  width: 1,
});

function clamp(value, min, max) {
  return Math.min(max, Math.max(min, value));
}

function coefficientForCutoff(frequency, sampleRate) {
  return 1 - Math.exp((-2 * Math.PI * frequency) / sampleRate);
}

export function greenLayerLevel(greenMix) {
  const mix = clamp(greenMix, 0, 1);
  return mix <= 0 ? 0 : mix ** 1.1 * 0.34;
}

export function createSleepToneChannelState(sampleRate = 48000) {
  return {
    brown: 0,
    previousWhite: 0,
    pinkB0: 0,
    pinkB1: 0,
    pinkB2: 0,
    pinkB3: 0,
    pinkB4: 0,
    pinkB5: 0,
    pinkB6: 0,
    air1: 0,
    air2: 0,
    rumble1: 0,
    rumble2: 0,
    greenLow: 0,
    greenFloor: 0,
    humPhase: Math.random() * Math.PI * 2,
    motionPhase: Math.random() * Math.PI * 2,
    flutterPhase: Math.random() * Math.PI * 2,
    drift: 0,
    driftTarget: 0,
    driftCounter: Math.floor(sampleRate * (0.35 + Math.random() * 0.4)),
    phaseOffset: Math.random() * Math.PI * 2,
    coefficientsReady: false,
    airCoeff1: 0,
    airCoeff2: 0,
    rumbleCoeff1: 0,
    rumbleCoeff2: 0,
    greenLowCoeff: 0,
    greenFloorCoeff: 0,
    sampleRate: 0,
  };
}

function ensureStateCoefficients(state, sampleRate) {
  if (state.coefficientsReady && state.sampleRate === sampleRate) {
    return;
  }

  state.sampleRate = sampleRate;
  state.airCoeff1 = coefficientForCutoff(700, sampleRate);
  state.airCoeff2 = coefficientForCutoff(420, sampleRate);
  state.rumbleCoeff1 = coefficientForCutoff(FAN_RUMBLE_FILTERS.firstCutoff, sampleRate);
  state.rumbleCoeff2 = coefficientForCutoff(FAN_RUMBLE_FILTERS.secondCutoff, sampleRate);
  state.greenLowCoeff = coefficientForCutoff(1800, sampleRate);
  state.greenFloorCoeff = coefficientForCutoff(220, sampleRate);
  state.coefficientsReady = true;
}

export function generatePinkSample(state, white) {
  state.pinkB0 = 0.99886 * state.pinkB0 + white * 0.0555179;
  state.pinkB1 = 0.99332 * state.pinkB1 + white * 0.0750759;
  state.pinkB2 = 0.969 * state.pinkB2 + white * 0.153852;
  state.pinkB3 = 0.8665 * state.pinkB3 + white * 0.3104856;
  state.pinkB4 = 0.55 * state.pinkB4 + white * 0.5329522;
  state.pinkB5 = -0.7616 * state.pinkB5 - white * 0.016898;

  const pink =
    state.pinkB0 +
    state.pinkB1 +
    state.pinkB2 +
    state.pinkB3 +
    state.pinkB4 +
    state.pinkB5 +
    state.pinkB6 +
    white * 0.5362;

  state.pinkB6 = white * 0.115926;

  return pink * 0.11;
}

export function generateBrownSample(state, white) {
  const brown = (state.brown + 0.02 * white) / 1.02;
  state.brown = brown;
  return brown * 3.5;
}

function generateGreenBandSample(state, white) {
  state.greenLow += state.greenLowCoeff * (white - state.greenLow);
  state.greenFloor += state.greenFloorCoeff * (white - state.greenFloor);
  return (state.greenLow - state.greenFloor) * 0.86;
}

export function generateSleepToneSample(state, white, config = DEFAULT_SLEEP_TONE_CONFIG, sampleRate = 48000) {
  ensureStateCoefficients(state, sampleRate);

  const pink = generatePinkSample(state, white);
  const brown = generateBrownSample(state, white);
  const fanAir = clamp(config.fanAir ?? DEFAULT_SLEEP_TONE_CONFIG.fanAir, 0, 1);
  const fanRumble = clamp(config.fanRumble ?? DEFAULT_SLEEP_TONE_CONFIG.fanRumble, 0, 1);
  const fanHum = clamp(config.fanHum ?? DEFAULT_SLEEP_TONE_CONFIG.fanHum, 0, 1);
  const fanHumPitch = config.fanHumPitch ?? DEFAULT_SLEEP_TONE_CONFIG.fanHumPitch;
  const fanDrift = clamp(config.fanDrift ?? DEFAULT_SLEEP_TONE_CONFIG.fanDrift, 0, 1);
  const warmth = clamp(config.warmth ?? DEFAULT_SLEEP_TONE_CONFIG.warmth, 0, 1);
  const greenLevel = greenLayerLevel(config.greenMix ?? DEFAULT_SLEEP_TONE_CONFIG.greenMix);

  const airSource = pink * (1 - FAN_AIR_BROWN_MIX) + brown * FAN_AIR_BROWN_MIX;
  state.air1 += state.airCoeff1 * (airSource - state.air1);
  state.air2 += state.airCoeff2 * (state.air1 - state.air2);
  const air = state.air2;

  const rumbleSource = brown * 0.9 + pink * 0.1;
  state.rumble1 += state.rumbleCoeff1 * (rumbleSource - state.rumble1);
  state.rumble2 += state.rumbleCoeff2 * (state.rumble1 - state.rumble2);
  const rumble = state.rumble2;

  state.driftCounter -= 1;
  if (state.driftCounter <= 0) {
    state.driftCounter = Math.floor(sampleRate * (0.35 + Math.random() * 0.65));
    state.driftTarget = Math.random() * 2 - 1;
  }

  state.drift += 0.00005 * (state.driftTarget - state.drift);
  state.motionPhase += (2 * Math.PI * (0.07 + fanDrift * 0.05)) / sampleRate;
  state.flutterPhase += (2 * Math.PI * (0.17 + fanDrift * 0.11)) / sampleRate;

  if (state.motionPhase > Math.PI * 2) {
    state.motionPhase -= Math.PI * 2;
  }

  if (state.flutterPhase > Math.PI * 2) {
    state.flutterPhase -= Math.PI * 2;
  }

  const motion =
    Math.sin(state.motionPhase + state.phaseOffset) * 0.6 +
    Math.sin(state.flutterPhase * 1.9 + state.phaseOffset * 0.37) * 0.4;
  const humFrequency = fanHumFrequency({
    fanHumPitch,
    fanDrift,
    drift: state.drift,
    motion,
  });

  state.humPhase += (2 * Math.PI * humFrequency) / sampleRate;
  if (state.humPhase > Math.PI * 2) {
    state.humPhase -= Math.PI * 2;
  }

  const humWave =
    Math.sin(state.humPhase) * 0.74 +
    Math.sin(state.humPhase * 2.01 + 0.4) * 0.18 +
    Math.sin(state.humPhase * 3.97 + 1.1) * 0.05;
  const bedMotion = 1 + fanDrift * (state.drift * 0.09 + motion * 0.06);
  const airLevel = fanAirLevel(fanAir) * (1 - warmth * 0.12);
  const rumbleLevel = fanRumbleLevel(fanRumble) * (1 + warmth * 0.08);
  const humLevel = fanHumLevel(fanHum);
  const fanBed = (air * airLevel + rumble * rumbleLevel) * bedMotion + humWave * humLevel;

  if (greenLevel <= 0) {
    return fanBed;
  }

  return fanBed + generateGreenBandSample(state, white) * greenLevel;
}

File: ios/SleepCompanionCore/Sources/SleepCompanionCore/SleepToneDSP.swift
import Foundation

public enum SleepToneDSP {
    public struct StereoSample: Equatable, Sendable {
        public var left: Double
        public var right: Double

        public init(left: Double, right: Double) {
            self.left = left
            self.right = right
        }
    }

    public struct ChannelRenderer: Sendable {
        private let sampleRate: Double
        private var random: SeededRandom
        private var leftState: ChannelState
        private var rightState: ChannelState

        public init(sampleRate: Double, seed: UInt64 = 0x5EED) {
            var random = SeededRandom(seed: seed)
            self.sampleRate = sampleRate
            self.random = random
            self.leftState = ChannelState(sampleRate: sampleRate, random: &random)
            self.rightState = ChannelState(sampleRate: sampleRate, random: &random)
        }

        public mutating func nextSample(parameters: SoundParameters) -> StereoSample {
            let parameters = parameters.clamped()
            let renderDrive = SoundOutputMapping.renderDrive(drive: parameters.drive)
            let leftWhite = random.nextSignedUnit()
            let rightWhite = random.nextSignedUnit()
            let left = Self.generateSample(
                state: &leftState,
                white: leftWhite,
                parameters: parameters,
                random: &random,
                sampleRate: sampleRate
            ) * renderDrive
            let right = Self.generateSample(
                state: &rightState,
                white: rightWhite,
                parameters: parameters,
                random: &random,
                sampleRate: sampleRate
            ) * renderDrive
            let widened = Self.applyStereoWidth(left: left, right: right, width: parameters.width)

            return StereoSample(left: Self.softLimited(widened.left), right: Self.softLimited(widened.right))
        }

        private static func generateSample(
            state: inout ChannelState,
            white: Double,
            parameters: SoundParameters,
            random: inout SeededRandom,
            sampleRate: Double
        ) -> Double {
            state.updateFilterCoefficientsIfNeeded(parameters: parameters, sampleRate: sampleRate)
            let pink = generatePinkSample(state: &state, white: white)
            let brown = generateBrownSample(state: &state, white: white)
            let greenLevel = SleepToneDSP.greenLayerLevel(parameters.greenMix)

            let airSource = pink * (1 - parameters.airTexture) + brown * parameters.airTexture
            state.air1 += state.airCoeff1 * (airSource - state.air1)
            state.air2 += state.airCoeff2 * (state.air1 - state.air2)

            let rumbleSource = brown * 0.9 + pink * 0.1
            state.rumble1 += state.rumbleCoeff1 * (rumbleSource - state.rumble1)
            state.rumble2 += state.rumbleCoeff2 * (state.rumble1 - state.rumble2)

            state.driftCounter -= 1
            if state.driftCounter <= 0 {
                state.driftCounter = Int(sampleRate * (0.35 + random.nextUnit() * 0.65))
                state.driftTarget = random.nextSignedUnit()
            }

            state.drift += 0.00005 * (state.driftTarget - state.drift)
            let movementRateScale = SoundFilterMapping.movementRateScale(speed: parameters.movementSpeed)
            state.motionPhase = wrapPhase(
                state.motionPhase + (2 * .pi * (0.07 + parameters.fanDrift * 0.05) * movementRateScale) / sampleRate
            )
            state.flutterPhase = wrapPhase(
                state.flutterPhase + (2 * .pi * (0.17 + parameters.fanDrift * 0.11) * movementRateScale) / sampleRate
            )

            let motion =
                sin(state.motionPhase + state.phaseOffset) * 0.6
                + sin(state.flutterPhase * 1.9 + state.phaseOffset * 0.37) * 0.4
            let humFrequency = max(
                30,
                parameters.fanHumPitch + (state.drift * 4.5 + motion * 2.2) * parameters.fanDrift
            )

            state.humPhase = wrapPhase(state.humPhase + (2 * .pi * humFrequency) / sampleRate)
            let harmonicScale = SoundFilterMapping.humHarmonicScale(harmonics: parameters.humHarmonics)
            let humWave =
                sin(state.humPhase) * 0.74
                + sin(state.humPhase * 2.01 + 0.4) * 0.18 * harmonicScale
                + sin(state.humPhase * 3.97 + 1.1) * 0.05 * harmonicScale

            let bedMotion = 1 + parameters.fanDrift * (state.drift * 0.09 + motion * 0.06)
            let airLevel = SleepToneDSP.fanAirLayerLevel(parameters.fanAir) * (1 - parameters.warmth * 0.12)
            let rumbleLevel = SleepToneDSP.fanRumbleLayerLevel(parameters.fanRumble) * (1 + parameters.warmth * 0.08)
            let humLevel = SleepToneDSP.fanHumLayerLevel(parameters.fanHum)
            let fanBed = (state.air2 * airLevel + state.rumble2 * rumbleLevel) * bedMotion + humWave * humLevel

            guard greenLevel > 0 else {
                return fanBed
            }

            return fanBed + generateGreenBandSample(state: &state, white: white) * greenLevel
        }

        private static func generatePinkSample(state: inout ChannelState, white: Double) -> Double {
            state.pinkB0 = 0.99886 * state.pinkB0 + white * 0.0555179
            state.pinkB1 = 0.99332 * state.pinkB1 + white * 0.0750759
            state.pinkB2 = 0.969 * state.pinkB2 + white * 0.153852
            state.pinkB3 = 0.8665 * state.pinkB3 + white * 0.3104856
            state.pinkB4 = 0.55 * state.pinkB4 + white * 0.5329522
            state.pinkB5 = -0.7616 * state.pinkB5 - white * 0.016898

            let pink =
                state.pinkB0
                + state.pinkB1
                + state.pinkB2
                + state.pinkB3
                + state.pinkB4
                + state.pinkB5
                + state.pinkB6
                + white * 0.5362

            state.pinkB6 = white * 0.115926
            return pink * 0.11
        }

        private static func generateBrownSample(state: inout ChannelState, white: Double) -> Double {
            let brown = (state.brown + 0.02 * white) / 1.02
            state.brown = brown
            return brown * 3.5
        }

        private static func generateGreenBandSample(state: inout ChannelState, white: Double) -> Double {
            state.greenLow += state.greenLowCoeff * (white - state.greenLow)
            state.greenFloor += state.greenFloorCoeff * (white - state.greenFloor)
            return (state.greenLow - state.greenFloor) * 0.86
        }

        private static func wrapPhase(_ phase: Double) -> Double {
            phase > 2 * .pi ? phase - 2 * .pi : phase
        }

        private static func applyStereoWidth(left: Double, right: Double, width: Double) -> StereoSample {
            let width = min(2, max(0, width))
            let mid = (left + right) * 0.5
            let side = (left - right) * 0.5 * width
            return StereoSample(left: mid + side, right: mid - side)
        }

        private static func softLimited(_ value: Double) -> Double {
            tanh(value)
        }
    }

    public static func greenLayerLevel(_ greenMix: Double) -> Double {
        let mix = min(1, max(0, greenMix))
        return mix <= 0 ? 0 : pow(mix, 1.7) * 0.12
    }

    static func fanAirLayerLevel(_ fanAir: Double) -> Double {
        pow(clamp(fanAir), 0.95) * 0.58
    }

    static func fanRumbleLayerLevel(_ fanRumble: Double) -> Double {
        pow(clamp(fanRumble), 1.15) * 0.78
    }

    static func fanHumLayerLevel(_ fanHum: Double) -> Double {
        pow(clamp(fanHum), 0.78) * 1.15
    }

    private static func clamp(_ value: Double) -> Double {
        min(1, max(0, value))
    }
}

private struct ChannelState: Sendable {
    var brown: Double = 0
    var pinkB0: Double = 0
    var pinkB1: Double = 0
    var pinkB2: Double = 0
    var pinkB3: Double = 0
    var pinkB4: Double = 0
    var pinkB5: Double = 0
    var pinkB6: Double = 0
    var air1: Double = 0
    var air2: Double = 0
    var rumble1: Double = 0
    var rumble2: Double = 0
    var greenLow: Double = 0
    var greenFloor: Double = 0
    var humPhase: Double
    var motionPhase: Double
    var flutterPhase: Double
    var drift: Double = 0
    var driftTarget: Double = 0
    var driftCounter: Int
    var phaseOffset: Double
    var airCoeff1: Double
    var airCoeff2: Double
    var rumbleCoeff1: Double
    var rumbleCoeff2: Double
    var greenLowCoeff: Double
    var greenFloorCoeff: Double
    var airColor: Double = 0.5
    var rumbleColor: Double = 0.5
    var greenColor: Double = 0.5

    init(sampleRate: Double, random: inout SeededRandom) {
        humPhase = random.nextUnit() * 2 * .pi
        motionPhase = random.nextUnit() * 2 * .pi
        flutterPhase = random.nextUnit() * 2 * .pi
        driftCounter = Int(sampleRate * (0.35 + random.nextUnit() * 0.4))
        phaseOffset = random.nextUnit() * 2 * .pi
        let airCutoffs = SoundFilterMapping.airCutoffs(color: airColor)
        let rumbleCutoffs = SoundFilterMapping.rumbleCutoffs(color: rumbleColor)
        let greenCutoffs = SoundFilterMapping.greenBandCutoffs(color: greenColor)
        airCoeff1 = Self.coefficient(forCutoff: airCutoffs.primary, sampleRate: sampleRate)
        airCoeff2 = Self.coefficient(forCutoff: airCutoffs.secondary, sampleRate: sampleRate)
        rumbleCoeff1 = Self.coefficient(forCutoff: rumbleCutoffs.primary, sampleRate: sampleRate)
        rumbleCoeff2 = Self.coefficient(forCutoff: rumbleCutoffs.secondary, sampleRate: sampleRate)
        greenLowCoeff = Self.coefficient(forCutoff: greenCutoffs.high, sampleRate: sampleRate)
        greenFloorCoeff = Self.coefficient(forCutoff: greenCutoffs.low, sampleRate: sampleRate)
    }

    mutating func updateFilterCoefficientsIfNeeded(parameters: SoundParameters, sampleRate: Double) {
        if airColor != parameters.airColor {
            airColor = parameters.airColor
            let cutoffs = SoundFilterMapping.airCutoffs(color: airColor)
            airCoeff1 = Self.coefficient(forCutoff: cutoffs.primary, sampleRate: sampleRate)
            airCoeff2 = Self.coefficient(forCutoff: cutoffs.secondary, sampleRate: sampleRate)
        }

        if rumbleColor != parameters.rumbleColor {
            rumbleColor = parameters.rumbleColor
            let cutoffs = SoundFilterMapping.rumbleCutoffs(color: rumbleColor)
            rumbleCoeff1 = Self.coefficient(forCutoff: cutoffs.primary, sampleRate: sampleRate)
            rumbleCoeff2 = Self.coefficient(forCutoff: cutoffs.secondary, sampleRate: sampleRate)
        }

        if greenColor != parameters.greenColor {
            greenColor = parameters.greenColor
            let cutoffs = SoundFilterMapping.greenBandCutoffs(color: greenColor)
            greenLowCoeff = Self.coefficient(forCutoff: cutoffs.high, sampleRate: sampleRate)
            greenFloorCoeff = Self.coefficient(forCutoff: cutoffs.low, sampleRate: sampleRate)
        }
    }

    private static func coefficient(forCutoff frequency: Double, sampleRate: Double) -> Double {
        1 - exp((-2 * .pi * frequency) / sampleRate)
    }
}

private struct SeededRandom: Sendable {
    private var state: UInt64

    init(seed: UInt64) {
        state = seed == 0 ? 0x5EED : seed
    }

    mutating func nextUnit() -> Double {
        state = state &* 6364136223846793005 &+ 1442695040888963407
        let bits = state >> 11
        return Double(bits) / Double(1 << 53)
    }

    mutating func nextSignedUnit() -> Double {
        nextUnit() * 2 - 1
    }
}
